File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJFunctionsIndex.java
package org.cappuccino_project.ide.intellij.plugin.indices;

import com.intellij.openapi.project.Project;
import com.intellij.psi.stubs.StubIndexKey;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Map;

public class ObjJFunctionsIndex extends ObjJStringStubIndexBase<ObjJFunctionDeclarationElement> {
    private static final StubIndexKey<String, ObjJFunctionDeclarationElement> KEY = IndexKeyUtil.createIndexKey(ObjJFunctionsIndex.class);
    private static final ObjJFunctionsIndex INSTANCE = new ObjJFunctionsIndex();
    private static final int VERSION = 0;


    private ObjJFunctionsIndex() {

    }

    public static ObjJFunctionsIndex getInstance()
    {
        return INSTANCE;
    }

    @Nullable
    public ObjJFunctionDeclarationElement get(@NotNull String fileName, @NotNull String functionName, @NotNull Project project) {
        for (ObjJFunctionDeclarationElement declarationElement : get(fileName, project)) {
            if (declarationElement.getFunctionNameAsString().equals(functionName)) {
                return declarationElement;
            }
        }
        return null;
    }

    @NotNull
    @Override
    public StubIndexKey<String, ObjJFunctionDeclarationElement> getKey() {
        return KEY;
    }

    @Override
    public int getVersion() {
        return super.getVersion() + VERSION;
    }

    @NotNull
    @Override
    protected Class<ObjJFunctionDeclarationElement> getIndexedElementClass() {
        return ObjJFunctionDeclarationElement.class;
    }
}

File: src/main/java/cappuccino/ide/intellij/plugin/indices/ObjJFunctionsIndex.kt
package cappuccino.ide.intellij.plugin.indices

import cappuccino.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement
import com.intellij.openapi.project.Project
import com.intellij.psi.stubs.StubIndexKey

class ObjJFunctionsIndex private constructor() : ObjJStringStubIndexBase<ObjJFunctionDeclarationElement<*>>() {

    override val indexedElementClass: Class<ObjJFunctionDeclarationElement<*>>
        get() = ObjJFunctionDeclarationElement::class.java

    operator fun get(fileName: String, functionName: String, project: Project): ObjJFunctionDeclarationElement<*>? {
        for (declarationElement in get(fileName, project)) {
            if (declarationElement.functionNameString == functionName) {
                return declarationElement
            }
        }
        return null
    }

    override fun getKey(): StubIndexKey<String, ObjJFunctionDeclarationElement<*>> {
        return KEY
    }

    override fun getVersion(): Int {
        return super.getVersion() + MINOR_VERSION + VERSION
    }

    companion object {
        private val KEY = IndexKeyUtil.createIndexKey<String, ObjJFunctionDeclarationElement<*>>(ObjJFunctionsIndex::class.java)
        val instance = ObjJFunctionsIndex()
        private const val MINOR_VERSION = 0
        private const val VERSION = 4
    }
}
